pkg/reconciler/route/resources/virtual_service.go

```diff
@@ -34,13 +37,40 @@ const (
 )
 
 // VirtualServiceName gets the name of a VirtualService given the route.
-func VirtualServiceName(hostname, domain, urlPath string) string {
-	return strconv.FormatUint(
+func VirtualServiceName(parts ...string) string {
+	prefix := strings.Join(parts, "-")
+
+	// Base 36 uses the characters 0-9a-z. The maximum number of chars it
+	// requires is 13 for a Uin64.
+	checksum := strconv.FormatUint(
 		crc64.Checksum(
-			[]byte(hostname+domain+path.Join("/", urlPath)),
+			[]byte(strings.Join(parts, "")),
 			crc64.MakeTable(crc64.ECMA),
 		),
-		10)
+		36)
+
+	prefix = strings.ToLower(prefix)
+
+	// Remove all non-alphanumeric characters.
+	prefix = regexp.MustCompile(`[^a-z0-9-_]`).ReplaceAllString(prefix, "-")
```

Comment:
Based on your above comment, is this necessary?

Reply:
The prefix may still have non-alphanumeric characters whereas the hash will only have 0-9a-z. We want to prepend the hostname and domain (and path in the case of a route). These things can contain illegal characters like a '/'.

Reply:
Sounds good. Can we move the prefixes to consts declared in this file?

Reply:
Well the prefix is generated based on what is passed in. So for example if the hostname is `some-hostname` and the domain is `example.com`, then we would end up with:

```
some-hostname-example-com-123456abcd
```

Reply:
Sorry, I meant `regexp.MustCompile(`[^a-z0-9-_]`)`

Reply:
Ah, got it.

Reply:
Done. Note, they have to be a `var` and not a `const`.